### PR TITLE
Enforce valid assignment target in the parser

### LINKED ISSUE #1929: An invalid code will cause an infinite loop
Hi all,

It seems that an invalid code will an infinite loop.
Ver.0.0.89 is OK, I guess.
Please see the following log.

```
$ cat a.js
(function() {
    "use strict";

    let x = function*() {
        yield ++(yield);
    };
}());
$ traceur --version
0.0.89$
$ traceur --out compile.js a.js
[ '/tmp/a.js:5:23: Unexpected token )',
  '/tmp/a.js:5:24: Unexpected token ;',
  '/tmp/a.js:6:5: Unexpected token }',
  '/tmp/a.js:6:6: Unexpected token ;',
  '/tmp/a.js:7:1: Unexpected token }',
  '/tmp/a.js:8:1: Unexpected token End of File',
  '/tmp/a.js:8:1: Unexpected token End of File',
  '/tmp/a.js:8:1: Unexpected token End of File' ]
$ ~/repository/traceur-compiler/traceur --version
0.0.90$
$ ~/repository/traceur-compiler/traceur --out compiled.js a.js
<<< No response >>>
```

### PR DESCRIPTION
**Pull request by arv**
This adds a function that matches the IsValidSimpleAssignmentTarget
spec algorithm

Fixes #1929

### CODE REVIEW

**Comment by arv** [comment]
@johnjbarton PTAL

File: src/syntax/Parser.js [hunk 1]
```diff
@@ -4505,6 +4514,14 @@ export class Parser {
   }
 
   reportReservedIdentifier_(token) {
-    this.reportError_(token, `${token.type} is a reserved identifier`);
+    this.reportError_(token.location, `${token.type} is a reserved identifier`);
```
Comment (arv) [inline src/syntax/Parser.js]: I can undo this change and make another PR that fixes all calls to always pass in a location.

File: src/staticsemantics/isValidSimpleAssignmentTarget.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * Matches "Static Semantics: IsValidSimpleAssignmentTarget" in the spec.
+ * @param {ParseTree} tree
+ * @return {boolean}
+ */
+export default function isValidSimpleAssignmentTarget(tree, isStrict) {
+  switch (tree.type) {
+    case IDENTIFIER_EXPRESSION: {
+      if (!isStrict) return true;
+      let {value} = tree.identifierToken;
+      return value !== 'arguments' && value !== 'eval';
```
Comment (johnjbarton) [inline src/staticsemantics/isValidSimpleAssignmentTarget.js]: Maybe `value` -> `token` or `identifier`?

**Comment by johnjbarton** [comment]
LGTM

File: src/staticsemantics/isValidSimpleAssignmentTarget.js (hunk 2 above)
Comment (arv) [inline src/staticsemantics/isValidSimpleAssignmentTarget.js]: I got lazy, ended up using value since it allows destructuring.

**Pull request merged**